Ensure shortcode renders for Privacy Policy & Terms of Service

lovo-h opened the pull request:
Fixes #7600

#### Changes proposed in this Pull Request

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

This PR allows the following shortcodes to be rendered in the WooPay Checkout preview: `[privacy_policy]` and `[terms]`. 

It should also be noted that the shortcodes have been renamed from `[privacy_policy_link]` to `[privacy_policy]` and `[terms_of_service_link]` to `[terms]`. These changes were made in order to have a consistent naming across WooPayments and WooCommerce.

https://github.com/Automattic/woocommerce-payments/assets/6844516/454f8673-557d-42f7-9bcf-37bc5b22533c


#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

**Test: Ensure the Privacy Policy and Terms of Service links appear in WooPay checkout**
* As a customer, navigate to the merchant shop, add an item to the cart, and proceed to WooPay checkout.
* Ensure that the **Policies and custom test** appear in the WooPay checkout page. 
* Ensure the Privacy Policy and Terms of Service links work as intended.

**Test: Ensure shortcode is rendered to look like a link, in the preview**
* As a merchant, ensure a Privacy Policy and Terms of Service page are linked.
* Navigate to **WooCommerce > Settings > Payments > WooPayments > WooPay > Checkout appearance**.
* Update the **Policies and custom text** so that it includes the following shortcodes: `[privacy_policy]` and `[terms]`.
* Ensure the shortcodes are rendered and appear as links. Also, ensure that they are not interactive/clickable.

**Regression test: Ensure other elements are not rendered in the preview**
* Update the **Policies and custom text** so that it includes other shortcodes, HTML elements, JS script, etc. 
* Ensure these other elements are not rendered.


-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

Commit:
Ensure shortcode renders for Privacy Policy & Terms of Service

Commit:
Add changelog

lovo-h commented:
@nikkivias, given that this is a preview, should merchants be able to click on the Privacy Policy and Terms of Service links and be navigated to the respective link? Or should it just appear like they are links but not be interactive?

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `7601` or branch name `fix/7600-woopay-checkout-shortcode-rendering` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=fix/7600-woopay-checkout-shortcode-rendering) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 9000f57eaacce0c948c36c269693942d5da7e807
- Build time: 2023-11-06 19:19:03 UTC

_Note: the build is updated when a new commit is pushed to this PR._

nikivias commented:
Hey @lovo-h good question. Because this is serves as a preview, we should remove any and all kind of interactions

Commit:
Shortcode links should not be interactive

Commit:
Rename shortcodes: [terms_of_service_link] and privacy_policy_link]

The shortcodes are renamed as follows:
* [terms_of_service_link] to [terms]
* [privacy_policy_link] to [privacy_policy]

cesarcosta99 commented on client/settings/express-checkout-settings/woopay-preview.js:
```diff
@@ -107,7 +108,36 @@ const CheckoutButton = ( { height } ) => {
 	const { style, ...restProps } = props;
-	const trimmedCustomMessage = ( customMessage || '' ).trim();
+
+	const preparedCustomMessage = useMemo( () => {
```
We may want to tweak this function because HTML elements are not being escaped in the preview:

<img width="692" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/10233985/5dba965a-7844-48f8-9f06-f40d8628ac3c">

Though, they're being escaped at the WooPay checkout 👍 

<img width="370" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/10233985/6df75a86-45d3-4e8c-9058-0e3362f3a9a3">

lovo-h replied:
We discussed this over Slack and agreed that these elements are being escaped, but in order to match the actual implementation, they should not appear in the preview at all. I agree that the preview should look as similar to the implementation as possible. Fixed in [0194e9e](https://github.com/Automattic/woocommerce-payments/pull/7601/commits/0194e9ef2e2266e61310585b25b8d9a107d13fa4) and [7fce67e](https://github.com/Automattic/woocommerce-payments/pull/7601/commits/7fce67ef64e5979e14180de2ef86c5ea1387874b).

cesarcosta99 reviewed (changes requested):
Code changes look good, and most of my tests passed.

- Test: Ensure the Privacy Policy and Terms of Service links appear in WooPay checkout ✅ 
- Test: Ensure shortcode is rendered to look like a link, in the preview ✅ 
- Regression test: Ensure other elements are not rendered in the preview ❌

Another issue I bumped into, that is unrelated to this PR, is when the ToS or Privacy Policy page is a draft (unpublished). At the checkout preview it looks OK, the text is rendered but not like a link. However, at the WooPay checkout, the link is rendered and lead to the correct permalink, but we got a 404 error. We may want to create an issue for it to figure what's the expected behavior should be.

frosso commented on includes/woopay/class-woopay-session.php:
```diff
@@ -633,10 +633,10 @@ private static function get_formatted_custom_message() {
 
 		$replacement_map = [
-			'[terms_of_service_link]' => wc_terms_and_conditions_page_id() ?
+			'[terms]'          => wc_terms_and_conditions_page_id() ?
```
How do these changes affect merchants that already saved their settings with the previous `[terms_of_service_link]` and `[privacy_policy_link]` values?

lovo-h replied:
Great observation. Added backwards compatibility in [58ba364](https://github.com/Automattic/woocommerce-payments/pull/7601/commits/58ba364cc8f263158ef55e7fc430b1a77f7d9459). 

I also added a mechanism to migrate these shortcodes, as they get updated.

Commit:
Backwards compatible shortcodes: [terms_of_service_link] & [privacy_policy_link]

In the WooPay checkout preview, add support for both
[terms_of_service_link] & [terms] and [privacy_policy_link] &
[privacy_policy]. All will appear as a link, in the preview.

In `update_woopay_custom_message(...)`, rename [terms_of_service_link]
to [terms] and [privacy_policy_link] to [privacy_policy].

In `get_formatted_custom_message()`, [terms_of_service_link] & [terms]
will both be replaced with a link to the "Terms of Service" and
[privacy_policy_link] & [privacy_policy] will both be replaced with a
link to the "Privacy Policy".

Commit:
Update WooPay checkout preview's "custom text" to match implementation

Commit:
Update WooPay checkout preview's "custom text" links to match implementation

lovo-h committed:
Merge branch 'develop' into fix/7600-woopay-checkout-shortcode-rendering

lovo-h committed:
Merge branch 'develop' into fix/7600-woopay-checkout-shortcode-rendering

cesarcosta99 commented on client/settings/express-checkout-settings/woopay-preview.js:
```diff
@@ -105,9 +109,62 @@ const CheckoutButton = ( { height } ) => {
+		if ( rawCustomMessage ) {
+			rawCustomMessage = sanitizeHtmlForPreview( rawCustomMessage );
+			rawCustomMessage = rawCustomMessage.replace(
+				/\[(terms|terms_of_service_link)\]/g,
+				'<span class="preview-layout__shortcode-link">Terms of Service</span>'
+			);
+			rawCustomMessage = rawCustomMessage.replace(
+				/\[(privacy_policy|privacy_policy_link)\]/g,
```
We can optimize the regular expressions like the following:

```suggestion
				/\[terms(_of_service_link)*\]/g,
				'<span class="preview-layout__shortcode-link">Terms of Service</span>'
			);
			rawCustomMessage = rawCustomMessage.replace(
				/\[privacy_policy(_link)*\]/g,
```

lovo-h replied:
Good call. I like that your version is less verbose. 

But in this case, having the entire term appear will make it easier to find when doing a global search, which I anticipate we will need to do in the future when we phase out the old shortcodes (e.g. `[terms_of_service_link]` and `[privacy_policy_link]`).

cesarcosta99 reviewed (approved):
Thanks for addressing the changes! LGTM 👍 

- Test: Ensure the Privacy Policy and Terms of Service links appear in WooPay checkout ✅ 
- Test: Ensure shortcode is rendered to look like a link, in the preview ✅ 
- Regression test: Ensure other elements are not rendered in the preview ✅ 

I left a tiny nitpick suggestion, but feel free to ignore it 😎
